homeassistant/components/homekit_controller/climate.py

```diff
@@ -56,15 +89,219 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+
+    @property
+    def max_temp(self):
+        """Return the maximum target temp."""
+        state = self.service.value(CharacteristicsTypes.TARGET_HEATER_COOLER_STATE)
+        if state == TargetHeaterCoolerStateValues.COOL and self.service.has(
+            CharacteristicsTypes.TEMPERATURE_COOLING_THRESHOLD
+        ):
+            return self.service[
+                CharacteristicsTypes.TEMPERATURE_COOLING_THRESHOLD
+            ].maxValue
+        if state == TargetHeaterCoolerStateValues.HEAT and self.service.has(
+            CharacteristicsTypes.TEMPERATURE_HEATING_THRESHOLD
+        ):
+            return self.service[
+                CharacteristicsTypes.TEMPERATURE_HEATING_THRESHOLD
+            ].maxValue
+        return super().max_temp
+
+    @property
+    def hvac_action(self):
+        """Return the current running hvac operation."""
+        # This characteristic describes the current mode of a device,
+        # e.g. a thermostat is "heating" a room to 75 degrees Fahrenheit.
+        # Can be 0 - 3 (Off, Idle, Heat, Cool)
+        value = self.service.value(CharacteristicsTypes.CURRENT_HEATER_COOLER_STATE)
+        return CURRENT_HEATER_COOLER_STATE_HOMEKIT_TO_HASS.get(value)
+
+    @property
+    def hvac_mode(self):
+        """Return hvac operation ie. heat, cool mode."""
+        # This characteristic describes the target mode
+        # E.g. should the device start heating a room if the temperature
+        # falls below the target temperature.
+        # CCan be 0 - 2 (Auto, Heat, Cool)
+        if (
+            self.service.has(CharacteristicsTypes.ACTIVE)
+            and self.service.value(CharacteristicsTypes.ACTIVE)
+            == ActivationStateValues.INACTIVE
+        ):
+            return HVAC_MODE_OFF
+        value = self.service.value(CharacteristicsTypes.TARGET_HEATER_COOLER_STATE)
+        return TARGET_HEATER_COOLER_STATE_HOMEKIT_TO_HASS.get(value)
+
+    @property
+    def hvac_modes(self):
+        """Return the list of available hvac operation modes."""
+        char = self.service[CharacteristicsTypes.TARGET_HEATER_COOLER_STATE]
+        ret = [
+            TARGET_HEATER_COOLER_STATE_HOMEKIT_TO_HASS[state]
+            for state in range(char.minValue, char.maxValue + 1)
+        ]
+        if self.service.has(CharacteristicsTypes.ACTIVE):
+            ret.append(HVAC_MODE_OFF)
+        return ret
+
+    @property
+    def swing_mode(self):
+        """Return the swing setting.
+
+        Requires SUPPORT_SWING_MODE.
+        """
+        value = self.service.value(CharacteristicsTypes.SWING_MODE)
+        return SWING_MODE_HOMEKIT_TO_HASS[value]
+
+    @property
+    def swing_modes(self):
+        """Return the list of available swing modes.
+
+        Requires SUPPORT_SWING_MODE.
+        """
+        char = self.service[CharacteristicsTypes.SWING_MODE]
+        return [
+            SWING_MODE_HOMEKIT_TO_HASS[swing]
+            for swing in range(char.minValue, char.maxValue + 1)
```

Comment:
You need to consider valid_values and not just minValue and maxValue. Manufactures can (and do) use both. clamp_enum_to_char may help here. There is an example in this file.